x/staking/keeper/grpc_query.go

```diff
@@ -0,0 +1,460 @@
+			"delegations for address %d don't exist", req.DelegatorAddr)
+	}
+	return &types.QueryDelegatorUnbondingDelegationsResponse{UnbondingResponses: unbondingDelegations, Res: res}, nil
+}
+
+// HistoricalInfo queries the historical info for given height
+func (k Querier) HistoricalInfo(c context.Context, req *types.QueryHistoricalInfoRequest) (*types.QueryHistoricalInfoResponse, error) {
+	if req == nil {
+		return nil, status.Errorf(codes.InvalidArgument, "empty request")
+	}
+
+	ctx := sdk.UnwrapSDKContext(c)
+	hi, found := k.GetHistoricalInfo(ctx, req.Height)
+	if !found {
+		return nil, status.Errorf(codes.NotFound, "historical info not found")
+	}
+
+	return &types.QueryHistoricalInfoResponse{Hist: &hi}, nil
+}
+
+func (k Querier) Redelegations(c context.Context, req *types.QueryRedelegationsRequest) (*types.QueryRedelegationsResponse, error) {
+	if req == nil {
+		return nil, status.Errorf(codes.InvalidArgument, "empty request")
+	}
+
+	var redels types.Redelegations
+	var res *query.PageResponse
+	var err error
+
+	ctx := sdk.UnwrapSDKContext(c)
+	store := ctx.KVStore(k.storeKey)
+	switch {
+	case !req.DelegatorAddr.Empty() && !req.SrcValidatorAddr.Empty() && !req.DstValidatorAddr.Empty():
+		redels, err = queryRedelegation(ctx, k, req)
+		if err != nil {
+			return nil, status.Errorf(
+				codes.NotFound,
+				"redelegation not found for delegator address %d", req.DelegatorAddr)
+		}
+	case req.DelegatorAddr.Empty() && !req.SrcValidatorAddr.Empty() && req.DstValidatorAddr.Empty():
+		redels, res, err = queryRedelegationsFromSrcValidator(store, k, req)
+	default:
+		redels, res, err = queryAllRedelegations(store, k, req)
+	}
+	if err != nil {
+		return nil, err
+	}
+
+	if len(redels) == 0 {
+		return nil, status.Errorf(
+			codes.NotFound,
+			"redelegation not found for given address %d from %d to %d",
+			req.DelegatorAddr, req.SrcValidatorAddr, req.DstValidatorAddr)
+	}
+	redelResponses, err := RedelegationsToRedelegationResponses(ctx, k.Keeper, redels)
+	if err != nil {
+		return nil, err
+	}
+
+	return &types.QueryRedelegationsResponse{RedelegationResponses: redelResponses, Res: res}, nil
+}
+
+func (k Querier) DelegatorValidators(c context.Context, req *types.QueryDelegatorValidatorsRequest) (*types.QueryDelegatorValidatorsResponse, error) {
+	if req == nil {
+		return nil, status.Errorf(codes.InvalidArgument, "empty request")
+	}
+
+	if req.DelegatorAddr.Empty() {
+		return nil, status.Errorf(codes.InvalidArgument, "invalid request")
+	}
+	var validators types.Validators
+	ctx := sdk.UnwrapSDKContext(c)
+
+	store := ctx.KVStore(k.storeKey)
+	delStore := prefix.NewStore(store, types.GetDelegationsKey(req.DelegatorAddr))
+	res, err := query.Paginate(delStore, req.Req, func(key []byte, value []byte) error {
+		delegation, err := types.UnmarshalDelegation(k.cdc, value)
+		if err != nil {
+			return err
+		}
+
+		validator, found := k.GetValidator(ctx, delegation.ValidatorAddress)
+		if !found {
+			return types.ErrNoValidatorFound
```

Comment:
use `status` error

Reply:
this is returned to paginate func, and later wrapped in status